spring-boot-project/spring-boot/src/main/java/org/springframework/boot/web/server/context/MissingWebServerFactoryBeanFailureAnalyzer.java

```diff
@@ -0,0 +1,124 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.web.server.context;
+
+import java.util.Arrays;
+
+import org.springframework.boot.WebApplicationType;
+import org.springframework.boot.diagnostics.AbstractFailureAnalyzer;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.boot.diagnostics.FailureAnalyzer;
+import org.springframework.context.ApplicationContextException;
+import org.springframework.util.ClassUtils;
+
+/**
+ * A {@link FailureAnalyzer} that performs analysis of failures caused by an
+ * {@link MissingWebServerFactoryBeanException}.
+ *
+ * @author Guirong Hu
+ */
+class MissingWebServerFactoryBeanFailureAnalyzer extends AbstractFailureAnalyzer<ApplicationContextException> {
+
+	private static final String SERVLET_CLASS = "jakarta.servlet.Servlet";
+
+	private static final String REACTIVE_WEB_SERVER_FACTORY_CLASS = "org.springframework.boot.web.reactive.server.ReactiveWebServerFactory";
+
+	private static final String[][] SERVLET_WEB_SERVER_FACTORY_CONDITIONAL_CLASSES = {
+			{ "org.apache.catalina.startup.Tomcat", "org.apache.coyote.UpgradeProtocol" },
+			{ "org.eclipse.jetty.server.Server", "org.eclipse.jetty.util.Loader",
+					"org.eclipse.jetty.webapp.Loader.WebAppContext" },
+			{ "io.undertow.Undertow" } };
+
+	private static final String[][] REACTIVE_WEB_SERVER_FACTORY_CONDITIONAL_CLASSES = {
+			{ "reactor.netty.http.server.HttpServer" }, { "org.apache.catalina.startup.Tomcat" },
+			{ "org.eclipse.jetty.server.Server", "org.eclipse.jetty.servlet.ServletHolder" },
+			{ "io.undertow.Undertow" } };
+
+	private final ClassLoader classLoader;
+
+	/**
+	 * Create a new {@code MissingWebServerFactoryBeanFailureAnalyzer} instance.
+	 */
+	MissingWebServerFactoryBeanFailureAnalyzer() {
+		this(null);
+	}
+
+	/**
+	 * Create a new {@code MissingWebServerFactoryBeanFailureAnalyzer} instance with the
+	 * given class loader.
+	 * @param classLoader the class loader
+	 */
+	MissingWebServerFactoryBeanFailureAnalyzer(ClassLoader classLoader) {
+		this.classLoader = classLoader;
+	}
+
+	@Override
+	protected FailureAnalysis analyze(Throwable rootFailure, ApplicationContextException cause) {
+		Throwable rootCause = cause.getCause();
+		if (rootCause instanceof MissingWebServerFactoryBeanException) {
+			WebApplicationType webApplicationType = getWebApplicationType(
+					((MissingWebServerFactoryBeanException) rootCause).getWebServerFactoryClass());
+			String action = action(webApplicationType);
+			if (action != null) {
+				return new FailureAnalysis(String.format(
+						"Reason: The running web application is of type %s, but the dependent class is missing.",
+						webApplicationType.name().toLowerCase()), action, cause);
+			}
+		}
+		return null;
+	}
+
+	private WebApplicationType getWebApplicationType(Class<?> webServerFactoryClass) {
```

Comment:
I think it was better when this was passed into the constructor of the exception. The thrower knows the type so there’s no need to make the analysis figure it out.

Reply:
Done.